### PR TITLE
totp doc

### PR DESCRIPTION
**Pull request by rgrullon**

### CODE REVIEW

File: docs/tutorials/enable-totp-openstack.md [hunk 1]
```diff
@@ -0,0 +1,93 @@
...
+
+Author: Ramon Grullon
+
+## Introduction
+
+Horizon added time-based one-time password (TOTP) authentication support,
+leveraging the already existing two factor authentication from Keystone.
+Now, if a user activates TOTP on Keystone, it gets activated on Horizon too.
+
+## Prerequesites
```
Comment (muertos) [inline docs/tutorials/enable-totp-openstack.md:11]: Should be Prerequisites

File: docs/tutorials/enable-totp-openstack.md [hunk 2]
```diff
@@ -0,0 +1,93 @@
...
+
+- A working OpenStack 2023.2 environment deployed using kolla-ansible.
+- Enabled totp on all control nodes
```
Comment (muertos) [inline docs/tutorials/enable-totp-openstack.md:14]: Ensure acronyms are in all caps

Comment (muertos) [inline docs/tutorials/enable-totp-openstack.md:14]: Should we explain how to run kolla-ansible against a cloud to enable TOTP in keystone? Maybe a mention of hey you need make this configuration change to keystone then run kolla-ansible with `--tags keystone` seems relevant to me.

File: docs/tutorials/enable-totp-openstack.md [hunk 3]
```diff
@@ -0,0 +1,93 @@
...
+- Admin access to the OpenStack dashboard and CLI.
+- Access to the kolla-ansible configuration files.
+
+### Create and Configure a TOTP Credential
+
+On one of your control nodes with OpenStack CLI configured:
+
+Create a project and user to utilize TOTP
```
Comment (muertos) [inline docs/tutorials/enable-totp-openstack.md:22]: A project is not necessarily needed to enable TOTP. I think explaining how to do this for an existing user makes sense. Additionally when a user is created with openstack cli, flags like `--enable-multi-factor-auth` can be set enabling this from the get-go

File: docs/tutorials/enable-totp-openstack.md [hunk 4]
```diff
@@ -0,0 +1,93 @@
...
+```
+
+TOTP uses a base32 encoded string for the secret. The secret must be at least
+128 bits (16 bytes). The following python code can be used to generate a TOTP secret:
+
+```python
+
+import base64
+message = b'1234567890123456'
+print(base64.b32encode(message).rstrip(b'='))
```
Comment (muertos) [inline docs/tutorials/enable-totp-openstack.md:39]: could also do this via bash `echo "iam16characters." | base32 | tr -d "="`

File: docs/tutorials/enable-totp-openstack.md [hunk 5]
```diff
@@ -0,0 +1,93 @@
...
+openstack credential create --type totp mfa-user GEZDGNBVGY3TQOJQGEZDGNBVGYZ
+
+openstack user set --multi-factor-auth-rule password,totp mfa-user
+
+openstack user set --enable-multi-factor-auth mfa-user
+```
+
+### Set Up TOTP
+
+This typically involves:
```
Comment (muertos) [inline docs/tutorials/enable-totp-openstack.md:61]: no need for the colon here

File: docs/tutorials/enable-totp-openstack.md [hunk 6]
```diff
@@ -0,0 +1,93 @@
...
+Scanning a QR code with a TOTP app like Google Authenticator.
+On a device install Google Authenticator and inside the app click on
+‘Set up account’ and then click on ‘Enter provided key’. In the input fields
+enter account name and secret.
+
+To quickly verify your configuration you can use <https://totp.app> and enter your
+secret to get your code.
+
+Verify TOTP Authentication
```
Comment (muertos) [inline docs/tutorials/enable-totp-openstack.md:70]: Maybe make this a numbered or bulleted list of steps to take

File: docs/tutorials/enable-totp-openstack.md [hunk 7]
```diff
@@ -0,0 +1,93 @@
...
+
+Log out of the OpenStack dashboard.
+
+Attempt to log back in. After entering your password, you should be prompted to
+enter your TOTP code.
+
+![totp](images/totp-login.png)
+
+![totp-success](images/totp-success.png)
```
Comment (muertos) [inline docs/tutorials/enable-totp-openstack.md:79]: I don't think there's much of a need to have a screenshot of logging back in to Horizon, I think just the box asking for the passcode should suffice.

File: docs/tutorials/enable-totp-openstack.md [hunk 8]
```diff
@@ -0,0 +1,93 @@
...
+
+Enter TOTP Code:
+
+Open your TOTP app and enter the current code to complete the login process.
+
+Ensure that administrators can manage users and their TOTP settings.
```
Comment (muertos) [inline docs/tutorials/enable-totp-openstack.md:85]: "Ensure that administrators can manage users and their TOTP settings." - how do we accomplish this? not too sure what this means

File: docs/tutorials/enable-totp-openstack.md (hunk 1 above)
Comment (rgrullon) [reply docs/tutorials/enable-totp-openstack.md:11]: fixed

File: docs/tutorials/enable-totp-openstack.md (hunk 2 above)
Comment (rgrullon) [reply docs/tutorials/enable-totp-openstack.md:14]: fixed

File: docs/tutorials/enable-totp-openstack.md [hunk 9]
```diff
@@ -0,0 +1,80 @@
...
+leveraging the already existing two factor authentication from Keystone.
+Now, if a user activates TOTP on Keystone, it gets activated on Horizon too.
+
+## Prerequisites
+
+- A working OpenStack 2023.2 environment deployed using kolla-ansible.
+- Enabled TOTP on all control nodes
+- Admin access to the OpenStack dashboard and CLI.
+- Access to the kolla-ansible configuration files.
+
```
Comment (muertos) [inline docs/tutorials/enable-totp-openstack.md:17]: Consider adding a section here, something like "### Reconfigure Keystone", explaining TOTP is not enabled by default and link to https://docs.openstack.org/keystone/latest/admin/auth-totp.html#configuring-totp, which explains the configuration needed to adjust Keystone. I can see a user following this guide and missing the keystone configuration step.

File: docs/tutorials/enable-totp-openstack.md [hunk 10]
```diff
@@ -0,0 +1,80 @@
...
+```
+
+Example output:
+
+```bash
+GEZDGNBVGY3TQOJQGEZDGNBVGY
+
+```
+
+### Configure credentials and mfa rules for user
```
Comment (muertos) [inline docs/tutorials/enable-totp-openstack.md:36]: Capitalize MFA since it is an acronym

File: docs/tutorials/enable-totp-openstack.md (hunk 9 above)
Comment (rgrullon) [reply docs/tutorials/enable-totp-openstack.md:17]: In v2 of the doc, I'll be creating a seperate section on how to configure keystone for TOTP.

Comment (muertos) [inline docs/tutorials/enable-totp-openstack.md:17]: To the Prerequisites, mention Keystone needing to be reconfigured, something like:
- [Keystone reconfigured with TOTP enabled](https://docs.openstack.org/keystone/latest/admin/auth-totp.html#configuring-totp)

File: docs/tutorials/enable-totp-openstack.md [hunk 11]
```diff
@@ -0,0 +1,80 @@
...
+- Enabled TOTP on all control nodes
+- Admin access to the OpenStack dashboard and CLI.
+- Access to the kolla-ansible configuration files.
+
+### Create and Configure a TOTP Credential
+
+On one of your control nodes with OpenStack CLI configured:
+
+TOTP uses a base32 encoded string for the secret. The secret must be at least
+128 bits (16 bytes). The following python code can be used to generate a TOTP secret:
```
Comment (muertos) [inline docs/tutorials/enable-totp-openstack.md:23]: this code snippet is now bash

File: docs/tutorials/enable-totp-openstack.md [hunk 12]
```diff
@@ -0,0 +1,80 @@
...
+openstack role add --user mfa-member --project admin member
+
+openstack credential create --type totp mfa-member GEZDGNBVGY3TQOJQGEZDGNBVGY
+
+```
+
+### Set Up TOTP
+
+This typically involves
+Scanning a QR code with a TOTP app like Google Authenticator.
```
Comment (muertos) [inline docs/tutorials/enable-totp-openstack.md:52]: "Scanning" should be lowercase

**Review by muertos — APPROVED** [review]

**Pull request merged**